engines/ic3ia.cpp

```diff
@@ -0,0 +1,369 @@
+      // then there's a (possibly abstract) counterexample
+      ProverResult refined = refine(pg);
+      if (refined == ProverResult::FALSE) {
+        // got a real counterexample trace
+        return false;
+      } else if (refined == ProverResult::UNKNOWN) {
+        // TODO: feed this through so it returns unknown
+        throw PonoException("Could not refine");
+      }
+    }
+  }
+  assert(!has_proof_goals());
+  return true;
+}
+
+bool IC3IA::intersects_bad()
+{
+  // we had to override this because ModelBasedIC3 adds the whole bad as a proof goal
+  // and here we just want the predicates
+  assert(reached_k_ + 2 == frames_.size());
+  assert(frame_labels_.size() == frames_.size());
+
+  push_solver_context();
+
+  // check if last frame intersects with bad
+  assert_frame_labels(reached_k_ + 1);
+  solver_->assert_formula(bad_);
+  Result r = solver_->check_sat();
+
+  if (r.is_sat()) {
+    add_proof_goal(get_conjunction_from_model(), reached_k_ + 1, NULL);
+  }
+
+  pop_solver_context();
+
+  return r.is_sat();
+}
+
+bool IC3IA::get_predecessor(size_t i,
+                            const Conjunction & c,
+                            Conjunction & out_pred)
+{
+  assert(i > 0);
+  assert(i < frames_.size());
+
+  assert(solver_context_ == 0);
+  push_solver_context();
+
+  // F[i-1]
+  assert_frame_labels(i - 1);
+  // -c
+  solver_->assert_formula(solver_->make_term(Not, c.term_));
+  // Trans
+  assert_trans_label();
+  // c'
+  solver_->assert_formula(next(c.term_));
+
+  Result r = solver_->check_sat();
+  if (r.is_sat()) {
+    // TODO: add generalize option
+    // TODO: refactor mbic3 to use get_conjunction_from_model and we won't need
+    // to override it
+    out_pred = get_conjunction_from_model();
+    pop_solver_context();
+  } else {
+    pop_solver_context();
+    assert(solver_context_ == 0);
+    TermVec assump, red_assump, rem_assump;
+    for (auto a : c.conjuncts_) {
+      assump.push_back(abs_ts_.next(a));
+    }
+
+    Term formula = make_and(TermVec{
+        get_frame(i - 1), solver_->make_term(Not, c.term_), trans_label_ });
+
+    // filter using unsatcore
+    reduce_assump_unsatcore(formula, assump, red_assump, &rem_assump);
+    // get current version of red_assump
+    TermVec cur_red_assump, cur_rem_assump;
+    for (auto a : red_assump) {
+      cur_red_assump.push_back(abs_ts_.curr(a));
+    }
+    for (auto a : rem_assump) {
+      cur_rem_assump.push_back(abs_ts_.curr(a));
+    }
+
+    fix_if_intersects_initial(cur_red_assump, cur_rem_assump);
+    assert(cur_red_assump.size() > 0);
+    out_pred = Conjunction(solver_, cur_red_assump);
+  }
+
+  assert(!r.is_unknown());
+  return r.is_sat();
+}
+
+Conjunction IC3IA::get_conjunction_from_model()
+{
+  TermVec conjuncts;
+  conjuncts.reserve(pred_statevars_.size());
+  Term val;
+  for (auto p : pred_statevars_) {
+    if ((val = solver_->get_value(p)) == true_) {
+      conjuncts.push_back(p);
+    } else {
+      assert(val == false_);
```

Comment:
isn't false depend on the solver? -- we use two different solvers in mbic3.

Reply:
Yeah there's a solver and an interpolator, the `false_` and `true_` here are for the main `solver_`, so this should be fine. Although, I see your point. Managing multiple solvers and their terms has been kind of a pain.

Reply:
We decided offline to use a naming schema to make this clearer. Implemented in 790a353ed8f90ad41f186af4e020be7402c6b490.